<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.10"/>
<title>My Project: /bioinfo/users/tmagalha/workspace/Release/src/genes.h File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">My Project
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.10 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
      <li><a href="globals.html"><span>File&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_68267d1309a1af8e8297ef4c3efbcdba.html">src</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">genes.h File Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>All functions to find gene of a read and treatment on gene file.  
<a href="#details">More...</a></p>
<div class="textblock"><code>#include &lt;stdio.h&gt;</code><br />
<code>#include &lt;stdlib.h&gt;</code><br />
<code>#include &lt;sys/stat.h&gt;</code><br />
<code>#include &lt;string.h&gt;</code><br />
<code>#include &quot;<a class="el" href="parser_8h_source.html">parser.h</a>&quot;</code><br />
</div>
<p><a href="genes_8h_source.html">Go to the source code of this file.</a></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structsize2__t.html">size2_t</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:a3b4633a5ddd8950da295fec64853d38d"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a3b4633a5ddd8950da295fec64853d38d"></a>
typedef struct <a class="el" href="structsize2__t.html">size2_t</a>&#160;</td><td class="memItemRight" valign="bottom"><b>size2_t</b></td></tr>
<tr class="separator:a3b4633a5ddd8950da295fec64853d38d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ac1f6db69cd42f249bba263c84186cc4a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsize2__t.html">size2_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="genes_8h.html#ac1f6db69cd42f249bba263c84186cc4a">getGeneAsNumber</a> (char *buffer_file, <a class="el" href="structRead.html">Read</a> *read)</td></tr>
<tr class="memdesc:ac1f6db69cd42f249bba263c84186cc4a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set gene for each read in "Read* read" with buffer_fil. Delete reads that didnt match.  <a href="#ac1f6db69cd42f249bba263c84186cc4a">More...</a><br /></td></tr>
<tr class="separator:ac1f6db69cd42f249bba263c84186cc4a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ef627c5c2834e8d76b75d0d717c4a36"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structsize2__t.html">size2_t</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="genes_8h.html#a5ef627c5c2834e8d76b75d0d717c4a36">getMGeneAsNumber</a> (char *buffer_file, <a class="el" href="structRead__chain.html">Read_chain</a> *backward)</td></tr>
<tr class="memdesc:a5ef627c5c2834e8d76b75d0d717c4a36"><td class="mdescLeft">&#160;</td><td class="mdescRight">Set mate gene for each backward in "Read_chain * backward" with buffer_file.  <a href="#a5ef627c5c2834e8d76b75d0d717c4a36">More...</a><br /></td></tr>
<tr class="separator:a5ef627c5c2834e8d76b75d0d717c4a36"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9860b8e98d6c3d4e714d30add5fbb46e"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="genes_8h.html#a9860b8e98d6c3d4e714d30add5fbb46e">getGeneLineBounds</a> (char *line, <a class="el" href="structsize2__t.html">size2_t</a> *size)</td></tr>
<tr class="memdesc:a9860b8e98d6c3d4e714d30add5fbb46e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Find bounds of gene in char * line.  <a href="#a9860b8e98d6c3d4e714d30add5fbb46e">More...</a><br /></td></tr>
<tr class="separator:a9860b8e98d6c3d4e714d30add5fbb46e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a972aa98e08bdf6d8f9634ae3a7e41a7a"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structRead__chain.html">Read_chain</a> **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="genes_8h.html#a972aa98e08bdf6d8f9634ae3a7e41a7a">getBackward</a> (<a class="el" href="structRead.html">Read</a> **reads, int nbchr, size_t *count)</td></tr>
<tr class="memdesc:a972aa98e08bdf6d8f9634ae3a7e41a7a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Gets every backward read from .  <a href="#a972aa98e08bdf6d8f9634ae3a7e41a7a">More...</a><br /></td></tr>
<tr class="separator:a972aa98e08bdf6d8f9634ae3a7e41a7a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6c345a0f8a6b0dc0fbddbeb36030fa0c"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="genes_8h.html#a6c345a0f8a6b0dc0fbddbeb36030fa0c">freeBackward</a> (<a class="el" href="structRead__chain.html">Read_chain</a> **chain, int nbchr)</td></tr>
<tr class="memdesc:a6c345a0f8a6b0dc0fbddbeb36030fa0c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Free each chromosome's backward list.  <a href="#a6c345a0f8a6b0dc0fbddbeb36030fa0c">More...</a><br /></td></tr>
<tr class="separator:a6c345a0f8a6b0dc0fbddbeb36030fa0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>All functions to find gene of a read and treatment on gene file. </p>
<dl class="section author"><dt>Author</dt><dd>Thomas MAGALHAES, Paul PANGANIBAN </dd></dl>
<dl class="section date"><dt>Date</dt><dd>August 26th 2015 </dd></dl>
</div><h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="a6c345a0f8a6b0dc0fbddbeb36030fa0c"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void freeBackward </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structRead__chain.html">Read_chain</a> **&#160;</td>
          <td class="paramname"><em>chain</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>nbchr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Free each chromosome's backward list. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">chain</td><td>the array of reads indexed by chr </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">nbchr</td><td>the number of chromosomes </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>void </dd></dl>

</div>
</div>
<a class="anchor" id="a972aa98e08bdf6d8f9634ae3a7e41a7a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structRead__chain.html">Read_chain</a>** getBackward </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structRead.html">Read</a> **&#160;</td>
          <td class="paramname"><em>reads</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>nbchr</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">size_t *&#160;</td>
          <td class="paramname"><em>count</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Gets every backward read from . </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir">[out]</td><td class="paramname">reads</td><td>The read array containing the entire reads </td></tr>
    <tr><td class="paramdir"></td><td class="paramname">nbchr</td><td>The number of chromosomes "reads" is made of </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">count</td><td>An array for counting the backwards / chromosomes </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A linked list of <a class="el" href="structRead__chain.html">Read_chain</a> of each backwards by chromosomes. </dd></dl>

</div>
</div>
<a class="anchor" id="ac1f6db69cd42f249bba263c84186cc4a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsize2__t.html">size2_t</a> getGeneAsNumber </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>buffer_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structRead.html">Read</a> *&#160;</td>
          <td class="paramname"><em>read</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set gene for each read in "Read* read" with buffer_fil. Delete reads that didnt match. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">buffer_file</td><td>Char * with gene ID </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">backward</td><td>Linked list of <a class="el" href="structRead.html">Read</a> that we will treat </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct*size2_t with attributes "a" equals to the number of reads which matched and attributes "b" the number of unmatched </dd></dl>

</div>
</div>
<a class="anchor" id="a9860b8e98d6c3d4e714d30add5fbb46e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t getGeneLineBounds </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>line</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structsize2__t.html">size2_t</a> *&#160;</td>
          <td class="paramname"><em>size</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Find bounds of gene in char * line. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">line</td><td>Contains the string from which the gene number should be extracted </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">size</td><td>Will contain the start and end bounds of the gene found. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>the gene value as a number </dd></dl>

</div>
</div>
<a class="anchor" id="a5ef627c5c2834e8d76b75d0d717c4a36"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structsize2__t.html">size2_t</a> getMGeneAsNumber </td>
          <td>(</td>
          <td class="paramtype">char *&#160;</td>
          <td class="paramname"><em>buffer_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structRead__chain.html">Read_chain</a> *&#160;</td>
          <td class="paramname"><em>backward</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Set mate gene for each backward in "Read_chain * backward" with buffer_file. </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramdir"></td><td class="paramname">buffer_file</td><td>Char * with gene ID </td></tr>
    <tr><td class="paramdir">[out]</td><td class="paramname">backward</td><td>Linked list of <a class="el" href="structRead__chain.html">Read_chain</a> that we will treat </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>a struct*size2_t with attributes "a" equals to the number of backwards which matched and attributes "b" the number of unmatched </dd></dl>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.10
</small></address>
</body>
</html>
